go/vt/srvtopo/server.go

```diff
@@ -47,3 +47,71 @@ type Server interface {
 	// a new value or an error occurs.
 	WatchSrvVSchema(ctx context.Context, cell string, callback func(*vschemapb.SrvVSchema, error))
 }
+
+// NewKeyspaceFilteringServer constructs a new server based on the provided
+// implementation that prevents the specified keyspaces from being exposed
+// to consumers of the new Server.
+func NewKeyspaceFilteringServer(underlying Server, selectedKeyspaces []string) Server {
+	keyspaces := map[string]bool{}
+	for _, ks := range selectedKeyspaces {
+		keyspaces[ks] = true
+	}
+
+	return keyspaceFilteringServer{
+		server:         underlying,
+		selectKeyspace: keyspaces,
+	}
+}
+
+type keyspaceFilteringServer struct {
+	server         Server
+	selectKeyspace map[string]bool
+}
+
+func (ksf keyspaceFilteringServer) GetTopoServer() *topo.Server {
+	return ksf.server.GetTopoServer()
```

Comment:
For safety you might want a nil check for ksf.server here.

Reply:
How would you feel about a compromise:

```go
func NewKeyspaceFilteringServer(
    underlying Server,
    selectedKeyspaces []string,
) (Server, error) {
    if underlying == nil {
        return nil, srvtopo.NoUnderlyingServer // or some other error
    }
    ...
}
```

My concern about checking `ksf.server` for nil is that it feels we should do that everywhere or nowhere. If we do it everywhere then, in the cases where we're making queries about the topology, I'm not sure what the least surprising response would be if there was no underlying server to contact... `topo.NoImplementation`? some new error code?

Happy to go either way, seems like a reasonable concern.

Reply:
Went with my suggestion returning `ErrNilUnderlyingServer` pending guidance on error states without an underlying server. The caveat is that it's not super difficult to trick an `== nil` check by something like:

```go
var underlyingPtr *someServerImpl
NewKeyspaceFilteringServer(underlyingPtr, []string{})
```

That said it also doesn't seem super likely and I'm not particularly worried about it.